src/pnpq/devices/refactored_odl_thorlabs_kbd101.py

```diff
@@ -0,0 +1,320 @@
+
+    def home(self) -> None:
+        self.set_channel_enabled(True)
+        start_time = time.perf_counter()
+        result = self.connection.send_message_expect_reply(
+            AptMessage_MGMSG_MOT_MOVE_HOME(
+                chan_ident=self._chan_ident,
+                destination=Address.GENERIC_USB,
+                source=Address.HOST_CONTROLLER,
+            ),
+            lambda message: (
+                isinstance(
+                    message,
+                    (
+                        AptMessage_MGMSG_MOT_MOVE_COMPLETED_20_BYTES,
+                        AptMessage_MGMSG_MOT_MOVE_STOPPED_20_BYTES,
+                    ),
+                )
+                and message.chan_ident == self._chan_ident
+                and message.destination == Address.HOST_CONTROLLER
+                and message.source == Address.GENERIC_USB
+            ),
+        )
+        # Sometimes the move stopped is received when interrupted
+        # by the user or when an invalid position is given
+        if isinstance(result, AptMessage_MGMSG_MOT_MOVE_STOPPED_20_BYTES):
+            self.log.warning(
+                "move_absolute command failed",
+                error="Move stopped before completion",
+            )
+            raise RuntimeError("Move stopped before completion")
+
+        elapsed_time = time.perf_counter() - start_time
+        self.log.debug("home command finished", elapsed_time=elapsed_time)
+        self.set_channel_enabled(False)
+
+    def move_absolute(self, position: Quantity) -> None:
+        absolute_distance = round(position.to("kbd101_position").magnitude)
+        self.set_channel_enabled(True)
+        self.log.debug("Sending move_absolute command...")
+        start_time = time.perf_counter()
+
+        result = self.connection.send_message_expect_reply(
+            AptMessage_MGMSG_MOT_MOVE_ABSOLUTE(
+                chan_ident=self._chan_ident,
+                absolute_distance=absolute_distance,
+                destination=Address.GENERIC_USB,
+                source=Address.HOST_CONTROLLER,
+            ),
+            lambda message: (
+                isinstance(
+                    message,
+                    (
+                        AptMessage_MGMSG_MOT_MOVE_COMPLETED_20_BYTES,
+                        AptMessage_MGMSG_MOT_MOVE_STOPPED_20_BYTES,
+                    ),
+                )
+                and message.chan_ident == self._chan_ident
+                and message.destination == Address.HOST_CONTROLLER
+                and message.source == Address.GENERIC_USB
+                and (  # If move is completed, check if the position is within 1mm of the target
+                    isinstance(message, AptMessage_MGMSG_MOT_MOVE_STOPPED_20_BYTES)
+                    or (
+                        message.position > absolute_distance - 1000
+                        and message.position < absolute_distance + 1000
+                    )
+                )
+            ),
+        )
```

Comment:
It might be best to move the position check outside of the lambda.

That way, we can raise an exception if the position isn't what we expect it to be.

I'm trying to think of a scenario when we could receive a MOVE_COMPLETED that isn't associated with the move command that we sent, but that shouldn't be possible.